python/tritonserver/_api/_model.py

```diff
@@ -180,40 +181,52 @@ def async_infer(
         if inference_request is None:
             inference_request = InferenceRequest(model=self, **kwargs)
 
-        if (inference_request.response_queue is not None) and (
-            not isinstance(inference_request.response_queue, asyncio.Queue)
-        ):
-            raise InvalidArgumentError(
-                "asyncio.Queue must be used for async response iterator"
-            )
-
+        # The 'request' is referencing pointers managed by the 'inference_request', so
+        # the 'inference_request' needs to be kept alive as long as the 'request' is
+        # alive.
         request = inference_request._create_tritonserver_inference_request()
 
-        response_iterator = AsyncResponseIterator(
-            self,
-            request,
-            inference_request.response_queue,
-            raise_on_error,
-        )
+        self._server.infer_async(request)
 
-        response_allocator = ResponseAllocator(
-            inference_request.output_memory_allocator,
-            inference_request.output_memory_type,
-        ).create_tritonserver_response_allocator()
+        class AsyncResponseIterator:
+            def __init__(self, model, request, inference_request):
+                self._model = model
+                self._request = request
+                self._inference_request = inference_request
+                self._complete = False
 
-        request.set_response_callback(
-            response_allocator, None, response_iterator._response_callback, None
-        )
+            def __aiter__(self):
+                return self
 
-        self._server.infer_async(request)
-        return response_iterator
+            async def __anext__(self):
+                if self._complete:
+                    raise StopAsyncIteration
+
+                # The binding could be improved by returning an awaitable object, but it
+                # is easier for now to pass in an async future object to be set by the
+                # binding when fetching responses.
+                future = asyncio.get_running_loop().create_future()
+                self._request.get_next_response(future)
+                response, flags = await future
+
+                response = InferenceResponse._from_tritonserver_inference_response(
+                    self._model,
+                    response,
+                    flags,
+                    self._inference_request.output_memory_type,
+                )
+                self._complete = response.final
+
+                return response
+
+        return AsyncResponseIterator(self, request, inference_request)
 
     def infer(
         self,
         inference_request: Optional[InferenceRequest] = None,
         raise_on_error: bool = True,
         **kwargs: Unpack[InferenceRequest],
-    ) -> ResponseIterator:
+    ) -> Iterable:
```

Comment:
why not use the explicit types?

Is it because of forward declaration?

If so we can use quotes, or import from typing future to allow for forward declaration,

```
"ResponseIterator"
```

Reply:
it is reverted to use the explicit type `ResponseIterator`.